add api design

ShutingQing opened the pull request:
#### What type of PR is this?

Add one of the following kinds:
* documentation


#### What this PR does / why we need it:
Add API Design.
Basically split "Get a Slice" function into 2 steps. One is reserve a slice. Another is manage the end service access control. 



#### Which issue(s) this PR fixes:

<!-- Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`. -->

Fixes #

#### Special notes for reviewers:



#### Changelog input

```
 release-note

```

#### Additional documentation 

This section can be blank.



```
docs

```

ShutingQing commented:
Please reviewers take a look in advance.

tlohmar commented on documentation/API_documentation/Network Slice Booking API Design.md:
```diff
@@ -0,0 +1,77 @@
+
+2. Scenario 2 - B2C
+
+OTT act as a Platform to Sell Slice
```
'OTT' typically not refers to an actor. What is meant here? Does OTT resolve to Over-The-Top?

Is this here an Application Provider or an Application Service Provider, who is the customer (not OTT)?

tlohmar replied:
How was this comment resolved?

tlohmar commented on documentation/API_documentation/Network Slice Booking API Design.md:
```diff
@@ -0,0 +1,77 @@
+
+Seller: Operator
+
+Mode: OTT only act as a platform who help provide slice to customer. 
```
I think, OPG uses the term 'Aggregator'.

tlohmar replied:
How was this comment resolved?

tlohmar commented on documentation/API_documentation/Network Slice Booking API Design.md:
```diff
@@ -0,0 +1,77 @@
+
+Step2: Individual purchase the slice from the OTT APP, pay to operators directly. 
+
+Step3: Provider provide the slice to the individuals (to C) directly. 
```
what is the relation between the 'Provider" and 'OTT'?

tlohmar commented on documentation/API_documentation/Network Slice Booking API Design.md:
```diff
@@ -0,0 +1,77 @@
+
+3. Scenario3 - B2B2C
+
+   OTT buy the slice, and resell the slice. 
```
The OTT refers here to an aggregator, who then resells the service to Enterprises or Consumers, correct?

ShutingQing replied:
OTT is enterprises. Resell to OTT toC customer. B to B to C. Operators to OTT to C End Customers.

tlohmar commented on documentation/API_documentation/Network Slice Booking API Design.md:
```diff
@@ -0,0 +1,77 @@
+
+##  **Procudure of How to Get a Slice:**
+
+General: Split the slice reservsation function and device access control function, to fit the above 3 scenario.
```
> reservsation 

typo

tlohmar commented on documentation/API_documentation/Network Slice Booking API Design.md:
```diff
@@ -0,0 +1,77 @@
+General: Split the slice reservsation function and device access control function, to fit the above 3 scenario.
+
+   - API 1：Reserve a slice resource 
+   - API 2：Manage the access control of devices to the slice
```
At what time is it possible to 'Manage access control'? 

In my understanding, the resource (a slice or a dedicated network) has at least two states, i.e. RESERVED, ACTIVE 
It should be possible to have device access control during both states.

tlohmar commented on documentation/API_documentation/Network Slice Booking API Design.md:
```diff
@@ -0,0 +1,77 @@
+   - API 2：Manage the access control of devices to the slice
+
+Basically there should be a CRUD for "Reserve a Slice Resource" and a CRUD for "Manage the access control of devices to the slice". 
+- API 1：Reserve a slice resource 
```
Which API should be used, while the reservation is 'active'? I think, API should not be named 'reserve'. Maybe 'manage'

tlohmar replied:
How is this comment resolved?

tlohmar commented on documentation/API_documentation/Network Slice Booking API Design.md:
```diff
@@ -0,0 +1,77 @@
+- API 1：Reserve a slice resource 
+- Function: 
+  - Reserve a slice of one selected period and one selected area. 
+  - Include End Devices as optional. 
+    - For those who would like to activate devices within 1 click, may provide the end devices infos. 
+    - For those who would like to later activate devices, or don't know which devices to active when order the slice resource, may later call API2 to do activation. Suitable for B2B2C scenario.
```
This can be simplified, since all the three bullets kind of say the same thing: Device Access Management may happen at any point in time while the resource is in revered-state and in active-state.

tlohmar commented on documentation/API_documentation/Network Slice Booking API Design.md:
```diff
@@ -0,0 +1,77 @@
+
+- API 2：Manage the access control of devices to the slice
+- Function:
+  - Slice Owners may have the right to allow access of devices to slice between the time of the order been made till the services ends. 
```
I am wondering about the 'may' in 'Slice Owners may have the right to allow access': What is the use-case, when the Slice Owner has NOT the right to allow access of devices? In which cases can the operator add other devices into the resource?

ShutingQing replied:
This point is where need to discuss, which design, is to mostly extent, make sense for developers.

1, Slice Owners, **is supposedly to have the right to** manage which devices to access the ordered Slice. 
2, To comply with different possible business scenarios, for scenario 2 and 3, OTT platform as the middle seller, when they purchase the slice resource, there are 2 situations. 1) OTT platform already know who will purchase slice service from them. So they can manage the end device access. 2) OTT platform not yet know, who will order slice service from them.

tlohmar replied:
> 2, To comply with different possible business scenarios, for scenario 2 and 3, OTT platform as the middle seller

Who is the Slice Owner in these Scenarios?

tlohmar reviewed (commented):
Generally, I am still not ok with the usage of the term 'slice', since it is against the CAMARA paradigm to hide telco complexity and allow the telco's to decide on the implementation.

ShutingQing commented:
> Generally, I am still not ok with the usage of the term 'slice', since it is against the CAMARA paradigm to hide telco complexity and allow the telco's to decide on the implementation

Hey Thortsen, understand your point. I'll kindly suggest not to discuss the term slice under each PR. https://github.com/camaraproject/NetworkSliceBooking/issues/12

tlohmar commented on documentation/API_documentation/SliceMode1.png:
I suggest to split the "Operators / Aggregator" box into two cases: 
(1) Operator <-< Enterprise Customer: Is this a B2B  
(2) Operator <-> Aggregator <-> Enterprise Customer: This is a B2B2X case

tlohmar commented on documentation/API_documentation/SliceMode2.png:
Also here, the Operator / Aggregator Box should be splitted

(1) Operator <-> OTT <-> toC Customer: Is this a B2B2X , here B2B2C  
(2) Operator <-> Aggregator <-> OTT <-> toC Customer: This is a B2B2B2C case 

When there is an aggregator, which entity is acting as a broker? OTT or also the aggregator?

tlohmar commented on documentation/API_documentation/SliceMode3.png:
Also here, the Operator / Aggregator Box should be splitted

(1) Operator <-> OTT <-> toC Customer: Is this a B2B2X , here B2B2C
(2) Operator <-> Aggregator <-> OTT <-> toC Customer: This is a B2B2B2C case

tlohmar reviewed (commented):
I am wondering, whether we need to differentiate between B2B2X and B2B2B2X from the API design perspective.

tlohmar reviewed (commented):
I have the impression, that several comments are marked as 'resolved' without doin any changes or comments.

ShutingQing commented:
> I have the impression, that several comments are marked as 'resolved' without doin any changes or comments.



> I have the impression, that several comments are marked as 'resolved' without doin any changes or comments.

Hi Thorsten, there are too many comments. Some of them I modified. Some of them that I don't think need changes or maybe didn't get your point, so it stay the same and I leave some comments. 

No worry, if you have any suggestions, feel free to review~ I'll handle that. This is still in progress stage. And better give me some suggested modifications. @tlohmar

ShutingQing commented:
One suggestions is @tlohmar . May I kindly suggest that, maybe you can split your questions and your review suggestions. : ) That will be easier for me do know which are the questions, and which are modification suggestions. Thanks in advance!

ShutingQing commented:
@tlohmar **I've listed all your comments here. Hopefully it can be more clear. Please check if there is any not included. I believe some of the questions are originated from “Mode” and “What is OTT/Aggregator" is not aligned. I will leave another comment below. After this is discussed clear, let's modify the left together.** 
 
**1. Question: toB Is this an abbreviation?**
-> : Yes. toB, toC means to business(eg enterprise customer), to customer(individual customer).

**2. Comment: What is a NaaS API. Please add a reverence.**
-> : Modified. NaaS API means CAMARA NaaS Service API.

**3. Comment: Line 16, I suggest to focus on this API, e.g. 'this API allows the booking of a slice / resource with very short lead times'.** 
-> : Modified.

**4. Question and Comment: Line20-23. 'OTT' typically not refers to an actor. What is meant here? Does OTT resolve to Over-The-Top? Is this here an Application Provider or an Application Service Provider, who is the customer (not OTT)?**

- [ ] -> : OTT not equals to Aggregator. This is mode question. Suggested to discuss together below. 

**5. Question and Comment: Line28-31. I think, OPG uses the term 'Aggregator'.**

- [ ] -> : OTT not equals to Aggregator. This is mode question. Suggested to discuss together below. 


**6. Question and Comment: Line 35-37. what is the relation between the 'Provider" and 'OTT'?**
-> : Modified. Provider means Telecom Operator.

**7. Question: Line 41-43. The OTT refers here to an aggregator, who then resells the service to Enterprises or Consumers, correct? Is this the B2B2B2C scenario? Operators 2 Aggregator 2 OTT 2 consumer? When correct, this is a scenario with two aggregators involved. I suggest to rename 'OTT' to 'Aggregator#2'**

- [ ] -> : This is mode question. Suggested to discuss together below. 


**8. Comment: Line57. Typo.**
-> : Modified

**9. Question: Line 57-60. At what time is it possible to 'Manage access control'?In my understanding, the resource (a slice or a dedicated network) has at least two states, i.e. RESERVED, ACTIVE. It should be possible to have device access control during both states.**
-> : Yes. 

**10. Comment: Line 60-63. Which API should be used, while the reservation is 'active'? I think, API should not be named 'reserve'. Maybe 'manage'**
-> : Yes.

**11. Comment: Line 66-68. This can be simplified, since all the three bullets kind of say the same thing: Device Access Management may happen at any point in time while the resource is in revered-state and in active-state.**
-> : Modifed. 

**12. Question: I am wondering about the 'may' in 'Slice Owners may have the right to allow access': What is the use-case, when the Slice Owner has NOT the right to allow access of devices? In which cases can the operator add other devices into the resource?** 

- [ ] -> : Mode question. Discuss below.

ShutingQing commented:
For the mode question:
1. OTT and Aggregator: I believe OTT is not Aggregator. OTT is cosumer from Aggregator. 

2. For the Mode, as Thorsten suggested,  B2B2B2C stands for Telecom Operator, to Aggregator, to toB OTT, to toC Customer. Logically this is correct. But I'm worrying this is to complex. How about sort NaaS Platform together, no matter the NaaS Platform is from Telecom Operator's NaaS Platform Portal, or it's from Aggregator's NaaS Platform Portal. Then the Mode is much clear. B2B2C. Stands for, B(CAMARA NaaS Service API) toB(OTT Enterprise) toC (End User End Customer). Any thoughts?

tlohmar commented:
Thanks for the elaboration.

> One suggestions is @tlohmar . May I kindly suggest that, maybe you can split your questions and your review suggestions

Hmm, I am making suggestions, e.g. "I suggest to split the "Operators / Aggregator" box into two cases". Still, the figures handle the "with aggregator" and "without aggregator" together.

tlohmar commented on documentation/API_documentation/SliceMode1.png:
Two suggestions:
1: Change "Enterprise Developer" to "Enterprise". Background: The Enterprise Developer is only interacting with the Service API during the development process. Once the development is completed, the API invoker functionality gets into production and the Developer is not involved anymore.

2: Create two figures: 
  Fig (1) is:  "Operator NaaS Platform" <->  "Enterprise"
  Fig (2) is:  "Operator NaaS Platform" <-> "Aggregator NaaS Platform" <->  "Enterprise"

tlohmar reviewed (commented):
Two suggestions embedded.

tlohmar reviewed (commented):
This _NaaS Portal_ Box_looks confusing: Does this picture mean, that one _NaaS Portal_ contains both, the operator and the aggregator portal in the same instance? 
Does the _NaaS Portal_ always contain both subcomponents (Operator NaaS Portal **and** Aggregator NaaS Portal) or only one (Operator NaaS Portal **or** Aggregator NaaS Portal)? 
I can make a suggestion, once I understand the intention.

Bart-Ericsson commented:
Existing material can be used https://github.com/camaraproject/WorkingGroups/blob/main/Marketing/documentation/MarketingMaterial/CAMARA%20Presentation.pptx
Please see slide 12 and 13

ShutingQing commented:
![CAMARA Community Given Architecture](https://github.com/camaraproject/NetworkSliceBooking/assets/103195423/3a8c7712-fd7b-4a55-852e-cdc258185a7d)


> Existing material can be used 

https://github.com/camaraproject/WorkingGroups/blob/main/Marketing/documentation/MarketingMaterial/CAMARA%20Presentation.pptx Please see slide 12 and 13

Bart-Ericsson commented:
> ![CAMARA Community Given Architecture](https://private-user-images.githubusercontent.com/103195423/333540552-3a8c7712-fd7b-4a55-852e-cdc258185a7d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY1NDgyMDIsIm5iZiI6MTcxNjU0NzkwMiwicGF0aCI6Ii8xMDMxOTU0MjMvMzMzNTQwNTUyLTNhOGM3NzEyLWZkN2ItNGE1NS04NTJlLWNkYzI1ODE4NWE3ZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyNFQxMDUxNDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yOGQxN2JkNTNhNmE2ODIzZGI1YjU5MzgzNTJjOWI5NWE4MjQzYmU1NWMxNGFmMDQ4N2RjNmEwNWQyNzFiNmEyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.tkC6hZfIqFvYuRTEtVuJFJJql2aFZFSGsd1nWncwiBU)
> 
> > Existing material can be used https://github.com/camaraproject/WorkingGroups/blob/main/Marketing/documentation/MarketingMaterial/CAMARA%20Presentation.pptx Please see slide 12 and 13
> 
> Thanks Bart . If every one align on this, i will modify the graph based on this.

Yes please, but also consider slide 13. As you can see the Service APIs extend into the operator environment upto the transformation function. The transformation function is outside the scope of Camara and every operator can define this transformation function in their own way to realize the intent of the Service API.

ShutingQing commented:
Yes totally agree. TF is out of the scope.

Masa8106 commented:
I hope this is not too late to comment.

There is text "The business modes of Network Slice Booking can be B2B, B2B2C.", which B2C has removed from.
I am not sure a reason why it has remove so far, however I believe B2C is a possible business mode which should be in scope.
Therefore, I would like to suggest B2C to be reverted.

Regarding figure illustration, I think we can change from "Enterprise" to "toB/toC Customer" in the figure of Mode 1.

What do you think about?

DanXu-ChinaTelecom commented:
From my side，this API is for developer,  B2C model refers to the end user as developer to use this API, maybe this situation doesn't exist for the time being.

DanXu-ChinaTelecom commented:
@ShutingQing Mode 2 and Model 3, I think the toC Customer will pay for app serivice with vip/svip experience supported by network slice from the OTT APPs(not OTT developer), the toC Customer indirectly pay for guaranteed network slice service.

Masa8106 commented:
@DanXu-ChinaTelecom: Thanks for your clarification. From perspective on how API is used by developers, I can make sense with what you mentioned that B2C model is not needed in this diagram.

@ShutingQing:  One suggestion. How about changing "toD OTT Developer" to "toD APP Developer" in Mode 2 and 3 of the diagram, because there is word "APP Developer" in the body text.

ShutingQing commented:
@Masa8106 Thanks Masaharu and Dan, good suggestion, let's do that.

ShutingQing commented:
> @ShutingQing Mode 2 and Model 3, I think the toC Customer will pay for app serivice with vip/svip experience supported by network slice from the OTT APPs(not OTT developer), the toC Customer indirectly pay for guaranteed network slice service.

Yes, that's true. For here it's both correct to use toD or toB to stand for OTT/APP developers. Just to see which one is more accurate or easy to understand. toD is more direct.

hdamker commented:
I'm a bit surprised about the business model discussions within this document. The scope of CAMARA is defined as "CAMARA only works on customer-facing northbound APIs." Customer in the sense of the end customer of the API, the term application service provider (ASP) is often used here in the meantime. Business model discussions are subject of the GSMA OGW business work stream, but do not need to be repeated here in CAMARA from my perspective. 

It would be good if you focus on how a customer who is consuming the Network Slice Booking API(s) would use it, independent of the type of customer. Within the picture from the CAMARA presentation (btw also part of the ProjectCharter) this are blue lines. How an aggregator (or "OTT" in your picture) will package it in own enriched products/service is out of scope.

I would also recommend to avoid telco specific terms like "OTT", this term does not make sense in a collaborative eco-system.

And on the formal side I suggest that the complete document belongs into supporting documents, the documentation folder is about documentation of the API. Also "Network Slice Booking API Introduction.pptx" need to be moved, within Documentation only markdown is expected, see ProjectStructureAndRoles (with ppt we have no version control and review process)).
